Add multidoc encrypt/decrypt for YAML sources

jar-o opened the pull request:
Hi this adds multidoc (or "streams") support for YAML files with multiple documents, e.g.

```
key: value
---
key: value
key2: value
```

It works and passes existing tests, but I wanted to get feedback before going too much further in case my approach needs revising. I still need to add tests and update docs.

The significant changes codewise I think are:

* I've added `Branches` (`[]TreeBranch`) to the `Tree` struct. If the document format understands and has "multidocs", this is used, otherwise it will use the existing `Branch` field -- even for single document YAML files.
* The `Emitter` and `Loader` interfaces have changed to accommodate `Branches`
* This required an update to `vendor/github.com/mozilla-services/yaml` and those sources have been updated (I made a previous PR to that repo already)

autrilla commented:
Thanks for taking the time to add this feature!

The approach looks good in general. I think it would be simpler to get rid of the `Branch` field altogether, and let stores handle multiple branches if they can, or they can error out otherwise. It's easy for stores to get the first branch anyway, and we could always add a helper function to the stores package to get the first branch or an error if there's more than one branch. I think that'd make the data types and emitter and loader interfaces simpler.

What do you think?

As a side note, this will only store the SOPS metadata in the first YAML document, right? That seems fine to me.

jar-o commented:
Ok, yeah ditching `Branch` makes sense. I'll update. Oh and I meant to mention that in my original comments, but you're right -- the SOPS metadata is computed for the physical file (i.e. `hash` etc) rather than each individual stream. It seemed simplest and didn't seem necessary to change.

jar-o commented:
ok, I've removed `tree.Branch` in favor of `tree.Branches`, updated docs, and added tests. please let me know if there are any other changes I should make.

autrilla commented on sops.go:
```diff
@@ -464,7 +491,7 @@ type EncryptedFileEmitter interface {
 // shown
 type PlainFileEmitter interface {
-	EmitPlainFile(TreeBranch) ([]byte, error)
+	EmitPlainFile(Tree) ([]byte, error)
```
I think this should still just take `[]TreeBranch` (or `TreeBranches`), as it doesn't use the metadata.

jar-o replied:
oh I see. ok

autrilla commented on sops.go:
```diff
@@ -284,50 +287,64 @@ func (branch TreeBranch) walkBranch(in TreeBranch, path []string, onLeaves func(
+	for _, branch := range tree.Branches {
+		err = walk(branch)
+		if err != nil {
+			break
```
We could just return the error here instead of breaking and checking the error again later.

autrilla commented on sops.go:
```diff
@@ -342,59 +359,69 @@ func (tree Tree) Decrypt(key []byte, cipher Cipher) (string, error) {
+	for _, branch := range tree.Branches {
+		err = walk(branch)
+		if err != nil {
+			break
```
Same as above

autrilla commented on stores/yaml/store.go:
```diff
@@ -118,45 +122,89 @@ func (store *Store) LoadEncryptedFile(in []byte) (sops.Tree, error) {
 	}
+
+	var branches sops.TreeBranches
+	for _, doc := range data {
```
I feel like either this or the previous for loop are redundant, no? We only store the metadata in the first document, and we're first removing the metadata entries from the first tree and then from *all* trees.

autrilla commented on stores/yaml/store.go:
```diff
@@ -118,45 +122,89 @@ func (store *Store) LoadEncryptedFile(in []byte) (sops.Tree, error) {
+			}
+			out = append(out, tout...)
+		}
+	} else {
```
Why is this `if ... else` needed? The loop you wrote would just emit the first branch and not add the  `---\n`, right? I don't think we need to special-case the 'single-branch' scenario with how you wrote the code.

jar-o replied:
ah looks like I just missed this in the refactor. thanks for the catch, I'll fix.

autrilla commented on stores/yaml/store.go:
```diff
@@ -118,45 +122,89 @@ func (store *Store) LoadEncryptedFile(in []byte) (sops.Tree, error) {
+			}
+			out = append(out[:], tmpout[:]...)
+		}
+	} else {
```
Same as above, I'm not sure why we need to handle the 'single-branch' case separately.

jar-o commented:
ok I believe I addressed all your comments. Also added back a section to docs for "YAML Streams". Please let me know if there is anything else missing or in need of straightening out. Thanks. :)

jar-o commented:
certainly! thanks for the guidance.
